File: ts/src/RsqSDK.ts
// Rsq Ts SDK

import { CategoryEntity } from './entity/CategoryEntity'
import { CountryOfAsylumEntity } from './entity/CountryOfAsylumEntity'
import { CountryOfOriginEntity } from './entity/CountryOfOriginEntity'
import { CountryOfResettlementEntity } from './entity/CountryOfResettlementEntity'
import { DemographicEntity } from './entity/DemographicEntity'
import { DepartureEntity } from './entity/DepartureEntity'
import { HelperEntity } from './entity/HelperEntity'
import { RegionEntity } from './entity/RegionEntity'
import { SubmissionEntity } from './entity/SubmissionEntity'
import { UrlFetchEntity } from './entity/UrlFetchEntity'
import { YearEntity } from './entity/YearEntity'

export type * from './RsqTypes'


import { inspect } from 'node:util'

import type { Context, Feature } from './types'

import { config } from './Config'
import { RsqEntityBase } from './RsqEntityBase'
import { Utility } from './utility/Utility'


import { BaseFeature } from './feature/base/BaseFeature'


const stdutil = new Utility()


class RsqSDK {
  _mode: string = 'live'
  _options: any
  _utility = new Utility()
  _features: Feature[]
  _rootctx: Context

  constructor(options?: any) {

    this._rootctx = this._utility.makeContext({
      client: this,
      utility: this._utility,
      config,
      options,
      shared: new WeakMap()
    })

    this._options = this._utility.makeOptions(this._rootctx)

    const struct = this._utility.struct
    const getpath = struct.getpath
    const items = struct.items

    if (true === getpath(this._options.feature, 'test.active')) {
      this._mode = 'test'
    }

    this._rootctx.options = this._options

    this._features = []

    const featureAdd = this._utility.featureAdd
    const featureInit = this._utility.featureInit

    items(this._options.feature, (fitem: [string, any]) => {
      const fname = fitem[0]
      const fopts = fitem[1]
      if (fopts.active) {
        featureAdd(this._rootctx, this._rootctx.config.makeFeature(fname))
      }
    })

    if (null != this._options.extend) {
      for (let f of this._options.extend) {
        featureAdd(this._rootctx, f)
      }
    }

    for (let f of this._features) {
      featureInit(this._rootctx, f)
    }

    const featureHook = this._utility.featureHook
    featureHook(this._rootctx, 'PostConstruct')
  }


  options() {
    return this._utility.struct.clone(this._options)
  }


  utility() {
    return this._utility.struct.clone(this._utility)
  }


  async prepare(fetchargs?: any) {
    const utility = this._utility
    const struct = utility.struct
    const clone = struct.clone

    const {
      makeContext,
      makeFetchDef,
      prepareHeaders,
      prepareAuth,
    } = utility

    fetchargs = fetchargs || {}

    let ctx: Context = makeContext({
      opname: 'prepare',
      ctrl: fetchargs.ctrl || {},
    }, this._rootctx)

    const options = this._options

    // Build spec directly from SDK options + user-provided fetch args.
    const spec: any = {
      base: options.base,
      prefix: options.prefix,
      suffix: options.suffix,
      path: fetchargs.path || '',
      method: fetchargs.method || 'GET',
      params: fetchargs.params || {},
      query: fetchargs.query || {},
      headers: prepareHeaders(ctx),
      body: fetchargs.body,
      step: 'start',
    }

    ctx.spec = spec

    // Merge user-provided headers over SDK defaults.
    if (fetchargs.headers) {
      const uheaders = fetchargs.headers
      for (let key in uheaders) {
        spec.headers[key] = uheaders[key]
      }
    }

    // Apply SDK auth (apikey, auth prefix, etc.)
    const authResult = prepareAuth(ctx)
    if (authResult instanceof Error) {
      return authResult
    }

    return makeFetchDef(ctx)
  }


  async direct(fetchargs?: any) {
    const utility = this._utility
    const fetcher = utility.fetcher
    const makeContext = utility.makeContext

    const fetchdef = await this.prepare(fetchargs)
    if (fetchdef instanceof Error) {
      return fetchdef
    }

    let ctx: Context = makeContext({
      opname: 'direct',
      ctrl: (fetchargs || {}).ctrl || {},
    }, this._rootctx)

    try {
      const fetched = await fetcher(ctx, fetchdef.url, fetchdef)

      if (null == fetched) {
        return { ok: false, err: ctx.error('direct_no_response', 'response: undefined') }
      }
      else if (fetched instanceof Error) {
        return { ok: false, err: fetched }
      }

      const status = fetched.status

      // No body responses (204 No Content, 304 Not Modified) and explicit
      // zero content-length must skip JSON parsing — fetched.json() would
      // throw `Unexpected end of JSON input` on an empty body.
      const headers = fetched.headers
      const contentLength = headers && 'function' === typeof headers.get
        ? headers.get('content-length')
        : (headers || {})['content-length']
      const noBody = 204 === status || 304 === status || '0' === String(contentLength)

      let json: any = undefined
      if (!noBody) {
        try {
          json = 'function' === typeof fetched.json ? await fetched.json() : fetched.json
        }
        catch (parseErr) {
          // Body wasn't valid JSON — surface the raw response rather than
          // throwing. data stays undefined; callers can inspect status/headers.
          json = undefined
        }
      }

      return {
        ok: status >= 200 && status < 300,
        status,
        headers: fetched.headers,
        data: json,
      }
    }
    catch (err: any) {
      return { ok: false, err }
    }
  }



  _category?: CategoryEntity

  // Idiomatic facade: `client.category.list()` / `client.category.load({ id })`.
  get category(): CategoryEntity {
    return (this._category ??= new CategoryEntity(this, undefined))
  }

  /** @deprecated Use `client.category` instead. */
  Category(data?: any) {
    const self = this
    return new CategoryEntity(self,data)
  }


  _country_of_asylum?: CountryOfAsylumEntity

  // Idiomatic facade: `client.country_of_asylum.list()` / `client.country_of_asylum.load({ id })`.
  get country_of_asylum(): CountryOfAsylumEntity {
    return (this._country_of_asylum ??= new CountryOfAsylumEntity(this, undefined))
  }

  /** @deprecated Use `client.country_of_asylum` instead. */
  CountryOfAsylum(data?: any) {
    const self = this
    return new CountryOfAsylumEntity(self,data)
  }


  _country_of_origin?: CountryOfOriginEntity

  // Idiomatic facade: `client.country_of_origin.list()` / `client.country_of_origin.load({ id })`.
  get country_of_origin(): CountryOfOriginEntity {
    return (this._country_of_origin ??= new CountryOfOriginEntity(this, undefined))
  }

  /** @deprecated Use `client.country_of_origin` instead. */
  CountryOfOrigin(data?: any) {
    const self = this
    return new CountryOfOriginEntity(self,data)
  }


  _country_of_resettlement?: CountryOfResettlementEntity

  // Idiomatic facade: `client.country_of_resettlement.list()` / `client.country_of_resettlement.load({ id })`.
  get country_of_resettlement(): CountryOfResettlementEntity {
    return (this._country_of_resettlement ??= new CountryOfResettlementEntity(this, undefined))
  }

  /** @deprecated Use `client.country_of_resettlement` instead. */
  CountryOfResettlement(data?: any) {
    const self = this
    return new CountryOfResettlementEntity(self,data)
  }


  _demographic?: DemographicEntity

  // Idiomatic facade: `client.demographic.list()` / `client.demographic.load({ id })`.
  get demographic(): DemographicEntity {
    return (this._demographic ??= new DemographicEntity(this, undefined))
  }

  /** @deprecated Use `client.demographic` instead. */
  Demographic(data?: any) {
    const self = this
    return new DemographicEntity(self,data)
  }


  _departure?: DepartureEntity

  // Idiomatic facade: `client.departure.list()` / `client.departure.load({ id })`.
  get departure(): DepartureEntity {
    return (this._departure ??= new DepartureEntity(this, undefined))
  }

  /** @deprecated Use `client.departure` instead. */
  Departure(data?: any) {
    const self = this
    return new DepartureEntity(self,data)
  }


  _helper?: HelperEntity

  // Idiomatic facade: `client.helper.list()` / `client.helper.load({ id })`.
  get helper(): HelperEntity {
    return (this._helper ??= new HelperEntity(this, undefined))
  }

  /** @deprecated Use `client.helper` instead. */
  Helper(data?: any) {
    const self = this
    return new HelperEntity(self,data)
  }


  _region?: RegionEntity

  // Idiomatic facade: `client.region.list()` / `client.region.load({ id })`.
  get region(): RegionEntity {
    return (this._region ??= new RegionEntity(this, undefined))
  }

  /** @deprecated Use `client.region` instead. */
  Region(data?: any) {
    const self = this
    return new RegionEntity(self,data)
  }


  _submission?: SubmissionEntity

  // Idiomatic facade: `client.submission.list()` / `client.submission.load({ id })`.
  get submission(): SubmissionEntity {
    return (this._submission ??= new SubmissionEntity(this, undefined))
  }

  /** @deprecated Use `client.submission` instead. */
  Submission(data?: any) {
    const self = this
    return new SubmissionEntity(self,data)
  }


  _url_fetch?: UrlFetchEntity

  // Idiomatic facade: `client.url_fetch.list()` / `client.url_fetch.load({ id })`.
  get url_fetch(): UrlFetchEntity {
    return (this._url_fetch ??= new UrlFetchEntity(this, undefined))
  }

  /** @deprecated Use `client.url_fetch` instead. */
  UrlFetch(data?: any) {
    const self = this
    return new UrlFetchEntity(self,data)
  }


  _year?: YearEntity

  // Idiomatic facade: `client.year.list()` / `client.year.load({ id })`.
  get year(): YearEntity {
    return (this._year ??= new YearEntity(this, undefined))
  }

  /** @deprecated Use `client.year` instead. */
  Year(data?: any) {
    const self = this
    return new YearEntity(self,data)
  }




  static test(testoptsarg?: any, sdkoptsarg?: any) {
    const struct = stdutil.struct
    const setpath = struct.setpath
    const getdef = struct.getdef
    const clone = struct.clone
    const setprop = struct.setprop

    const sdkopts = getdef(clone(sdkoptsarg), {})
    const testopts = getdef(clone(testoptsarg), {})
    setprop(testopts, 'active', true)
    setpath(sdkopts, 'feature.test', testopts)

    const testsdk = new RsqSDK(sdkopts)
    testsdk._mode = 'test'

    return testsdk
  }


  tester(testopts?: any, sdkopts?: any) {
    return RsqSDK.test(testopts, sdkopts)
  }


  toJSON() {
    return { name: 'Rsq' }
  }

  toString() {
    return 'Rsq ' + this._utility.struct.jsonify(this.toJSON())
  }

  [inspect.custom]() {
    return this.toString()
  }

}




const SDK = RsqSDK


export {
  stdutil,

  BaseFeature,
  RsqEntityBase,

  RsqSDK,
  SDK,
}

File: go/core/rsq_sdk.go
package core

import (
	"fmt"

	vs "github.com/voxgig-sdk/rsq-sdk/go/utility/struct"
)

type RsqSDK struct {
	Mode     string
	options  map[string]any
	utility  *Utility
	Features []Feature
	rootctx  *Context
}

func NewRsqSDK(options map[string]any) *RsqSDK {
	sdk := &RsqSDK{
		Mode:     "live",
		Features: []Feature{},
	}

	sdk.utility = NewUtility()

	config := MakeConfig()

	sdk.rootctx = sdk.utility.MakeContext(map[string]any{
		"client":  sdk,
		"utility": sdk.utility,
		"config":  config,
		"options": options,
		"shared":  map[string]any{},
	}, nil)

	sdk.options = sdk.utility.MakeOptions(sdk.rootctx)

	if vs.GetPath([]any{"feature", "test", "active"}, sdk.options) == true {
		sdk.Mode = "test"
	}

	sdk.rootctx.Options = sdk.options

	// Add features in the resolved order (MakeOptions puts an explicit array
	// order first, else defaults to test-first). Ordering matters: the `test`
	// feature installs the base mock transport and the transport features
	// (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
	// must be added before them to sit at the base of the chain.
	featureOpts := ToMapAny(vs.GetProp(sdk.options, "feature"))
	if featureOpts != nil {
		if fo, ok := vs.GetPath([]any{"__derived__", "featureorder"}, sdk.options).([]any); ok {
			for _, n := range fo {
				fname, _ := n.(string)
				fopts := ToMapAny(featureOpts[fname])
				if fopts != nil {
					if active, ok := fopts["active"]; ok {
						if ab, ok := active.(bool); ok && ab {
							sdk.utility.FeatureAdd(sdk.rootctx, makeFeature(fname))
						}
					}
				}
			}
		}
	}

	// Add extension features.
	if extend := vs.GetProp(sdk.options, "extend"); extend != nil {
		if extList, ok := extend.([]any); ok {
			for _, f := range extList {
				if feat, ok := f.(Feature); ok {
					sdk.utility.FeatureAdd(sdk.rootctx, feat)
				}
			}
		}
	}

	// Initialize features.
	for _, f := range sdk.Features {
		sdk.utility.FeatureInit(sdk.rootctx, f)
	}

	sdk.utility.FeatureHook(sdk.rootctx, "PostConstruct")

	return sdk
}

func (sdk *RsqSDK) OptionsMap() map[string]any {
	out := vs.Clone(sdk.options)
	if om, ok := out.(map[string]any); ok {
		return om
	}
	return map[string]any{}
}

func (sdk *RsqSDK) GetUtility() *Utility {
	return CopyUtility(sdk.utility)
}

func (sdk *RsqSDK) GetRootCtx() *Context {
	return sdk.rootctx
}

func (sdk *RsqSDK) Prepare(fetchargs map[string]any) (map[string]any, error) {
	utility := sdk.utility

	if fetchargs == nil {
		fetchargs = map[string]any{}
	}

	var ctrl map[string]any
	if c := vs.GetProp(fetchargs, "ctrl"); c != nil {
		if cm, ok := c.(map[string]any); ok {
			ctrl = cm
		}
	}
	if ctrl == nil {
		ctrl = map[string]any{}
	}

	ctx := utility.MakeContext(map[string]any{
		"opname": "prepare",
		"ctrl":   ctrl,
	}, sdk.rootctx)

	options := sdk.options

	path, _ := vs.GetProp(fetchargs, "path").(string)
	method, _ := vs.GetProp(fetchargs, "method").(string)
	if method == "" {
		method = "GET"
	}

	params := ToMapAny(vs.GetProp(fetchargs, "params"))
	if params == nil {
		params = map[string]any{}
	}
	query := ToMapAny(vs.GetProp(fetchargs, "query"))
	if query == nil {
		query = map[string]any{}
	}

	headers := utility.PrepareHeaders(ctx)

	base, _ := vs.GetProp(options, "base").(string)
	prefix, _ := vs.GetProp(options, "prefix").(string)
	suffix, _ := vs.GetProp(options, "suffix").(string)

	ctx.Spec = NewSpec(map[string]any{
		"base":    base,
		"prefix":  prefix,
		"suffix":  suffix,
		"path":    path,
		"method":  method,
		"params":  params,
		"query":   query,
		"headers": headers,
		"body":    vs.GetProp(fetchargs, "body"),
		"step":    "start",
	})

	// Merge user-provided headers.
	if uh := vs.GetProp(fetchargs, "headers"); uh != nil {
		if uhm, ok := uh.(map[string]any); ok {
			for k, v := range uhm {
				ctx.Spec.Headers[k] = v
			}
		}
	}

	_, err := utility.PrepareAuth(ctx)
	if err != nil {
		return nil, err
	}

	return utility.MakeFetchDef(ctx)
}

func (sdk *RsqSDK) Direct(fetchargs map[string]any) (map[string]any, error) {
	utility := sdk.utility

	fetchdef, err := sdk.Prepare(fetchargs)
	if err != nil {
		return map[string]any{"ok": false, "err": err}, nil
	}

	if fetchargs == nil {
		fetchargs = map[string]any{}
	}

	var ctrl map[string]any
	if c := vs.GetProp(fetchargs, "ctrl"); c != nil {
		if cm, ok := c.(map[string]any); ok {
			ctrl = cm
		}
	}
	if ctrl == nil {
		ctrl = map[string]any{}
	}

	ctx := utility.MakeContext(map[string]any{
		"opname": "direct",
		"ctrl":   ctrl,
	}, sdk.rootctx)

	url, _ := fetchdef["url"].(string)
	fetched, fetchErr := utility.Fetcher(ctx, url, fetchdef)

	if fetchErr != nil {
		return map[string]any{"ok": false, "err": fetchErr}, nil
	}

	if fetched == nil {
		return map[string]any{
			"ok":  false,
			"err": ctx.MakeError("direct_no_response", "response: undefined"),
		}, nil
	}

	if fm, ok := fetched.(map[string]any); ok {
		status := ToInt(vs.GetProp(fm, "status"))
		headers := vs.GetProp(fm, "headers")

		// No-body responses (204, 304) and explicit zero content-length
		// must skip JSON parsing — calling json() on an empty body errors.
		var contentLength string
		if hm, ok := headers.(map[string]any); ok {
			if cl, ok := hm["content-length"]; ok {
				contentLength = fmt.Sprintf("%v", cl)
			}
		}
		noBody := status == 204 || status == 304 || contentLength == "0"

		var jsonData any
		if !noBody {
			if jf := vs.GetProp(fm, "json"); jf != nil {
				if f, ok := jf.(func() any); ok {
					// f() returns nil on parse error in our fetcher.
					jsonData = f()
				}
			}
		}

		return map[string]any{
			"ok":      status >= 200 && status < 300,
			"status":  status,
			"headers": headers,
			"data":    jsonData,
		}, nil
	}

	return map[string]any{"ok": false, "err": ctx.MakeError("direct_invalid", "invalid response type")}, nil
}


// Category returns a Category entity bound to this client.
// Idiomatic usage: client.Category(nil).List(nil, nil) or
// client.Category(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *RsqSDK) Category(data map[string]any) RsqEntity {
	return NewCategoryEntityFunc(sdk, data)
}


// CountryOfAsylum returns a CountryOfAsylum entity bound to this client.
// Idiomatic usage: client.CountryOfAsylum(nil).List(nil, nil) or
// client.CountryOfAsylum(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *RsqSDK) CountryOfAsylum(data map[string]any) RsqEntity {
	return NewCountryOfAsylumEntityFunc(sdk, data)
}


// CountryOfOrigin returns a CountryOfOrigin entity bound to this client.
// Idiomatic usage: client.CountryOfOrigin(nil).List(nil, nil) or
// client.CountryOfOrigin(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *RsqSDK) CountryOfOrigin(data map[string]any) RsqEntity {
	return NewCountryOfOriginEntityFunc(sdk, data)
}


// CountryOfResettlement returns a CountryOfResettlement entity bound to this client.
// Idiomatic usage: client.CountryOfResettlement(nil).List(nil, nil) or
// client.CountryOfResettlement(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *RsqSDK) CountryOfResettlement(data map[string]any) RsqEntity {
	return NewCountryOfResettlementEntityFunc(sdk, data)
}


// Demographic returns a Demographic entity bound to this client.
// Idiomatic usage: client.Demographic(nil).List(nil, nil) or
// client.Demographic(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *RsqSDK) Demographic(data map[string]any) RsqEntity {
	return NewDemographicEntityFunc(sdk, data)
}


// Departure returns a Departure entity bound to this client.
// Idiomatic usage: client.Departure(nil).List(nil, nil) or
// client.Departure(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *RsqSDK) Departure(data map[string]any) RsqEntity {
	return NewDepartureEntityFunc(sdk, data)
}


// Helper returns a Helper entity bound to this client.
// Idiomatic usage: client.Helper(nil).List(nil, nil) or
// client.Helper(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *RsqSDK) Helper(data map[string]any) RsqEntity {
	return NewHelperEntityFunc(sdk, data)
}


// Region returns a Region entity bound to this client.
// Idiomatic usage: client.Region(nil).List(nil, nil) or
// client.Region(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *RsqSDK) Region(data map[string]any) RsqEntity {
	return NewRegionEntityFunc(sdk, data)
}


// Submission returns a Submission entity bound to this client.
// Idiomatic usage: client.Submission(nil).List(nil, nil) or
// client.Submission(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *RsqSDK) Submission(data map[string]any) RsqEntity {
	return NewSubmissionEntityFunc(sdk, data)
}


// UrlFetch returns a UrlFetch entity bound to this client.
// Idiomatic usage: client.UrlFetch(nil).List(nil, nil) or
// client.UrlFetch(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *RsqSDK) UrlFetch(data map[string]any) RsqEntity {
	return NewUrlFetchEntityFunc(sdk, data)
}


// Year returns a Year entity bound to this client.
// Idiomatic usage: client.Year(nil).List(nil, nil) or
// client.Year(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *RsqSDK) Year(data map[string]any) RsqEntity {
	return NewYearEntityFunc(sdk, data)
}



func TestSDK(testopts map[string]any, sdkopts map[string]any) *RsqSDK {
	if sdkopts == nil {
		sdkopts = map[string]any{}
	}
	sdkopts = vs.Clone(sdkopts).(map[string]any)

	if testopts == nil {
		testopts = map[string]any{}
	}
	testopts = vs.Clone(testopts).(map[string]any)
	testopts["active"] = true

	vs.SetPath(sdkopts, []any{"feature", "test"}, testopts)

	sdk := NewRsqSDK(sdkopts)
	sdk.Mode = "test"

	return sdk
}
